### PR TITLE
prompt user to update go.buildTags if file has the build tags

### PR DESCRIPTION
**Pull request by urohit011**
This one is for #1497, I have added check to prompt the user to update "go.buildTags" settings if the file has build tags that are not specified in "go.buildTags". The prompt has two options, "Don't show again" and "Open settings".

### CODE REVIEW

**Comment by urohit011** [comment]
@ramya-rao-a will you please take a look at this, I really don't understand why the tests are failing.

**Comment by urohit011** [comment]
Thanks for your observation @ramya-rao-a, and sorry for late reply. I'll make the changes asap.

File: src/goCheck.ts [hunk 1]
```diff
@@ -45,6 +45,34 @@ export function notifyIfGeneratedFile(e: vscode.TextDocumentChangeEvent) {
 			}
 		});
 	}
+
+	let settingsBuildTags = goConfig['buildTags'] ? goConfig['buildTags'].split(' ') : null;
+	let fileBuildTags = [];
+	const openSettings = {title: 'Open settings'};
+
+	for (let i = 0; i < e.document.lineCount; i++) {
```
[the remark this answers is not in the document]
Comment (urohit011) [reply src/goCheck.ts]: You were right, we did not have to loop through all the lines, first I thought of breaking out of the loop after encountering a "package" statement, but empty line is fine.

File: src/goCheck.ts [hunk 2]
```diff
@@ -45,6 +45,34 @@ export function notifyIfGeneratedFile(e: vscode.TextDocumentChangeEvent) {
...
+		if (e.document.lineAt(i).text.match(/^\/\/ \+build [\w\s]+$/)) {
```
[the remark this answers is not in the document]
Comment (urohit011) [reply src/goCheck.ts]: I could not find the list of all the build tags on internet that is why I have used "." to match the build tags.

File: src/goCheck.ts [hunk 3]
```diff
@@ -45,6 +45,34 @@ export function notifyIfGeneratedFile(e: vscode.TextDocumentChangeEvent) {
...
+			fileBuildTags.push(...e.document.lineAt(i).text.split(' '));
+		}
+	}
+
+	fileBuildTags = fileBuildTags.filter(tag => tag !== '+build' && tag !== '//' && tag !== '' && tag !== '//+build');
```
[the remark this answers is not in the document]
Comment (urohit011) [reply src/goCheck.ts]: I could not get rid of the filter totally, but have reduced it. What do you suggest?

**Comment by urohit011** [comment]
Hello @ramya-rao-a I have made a couple of changes. Will you please take a look at it?

File: src/goCheck.ts [hunk 4]
```diff
@@ -45,6 +45,37 @@ export function notifyIfGeneratedFile(e: vscode.TextDocumentChangeEvent) {
...
+
+	let settingsBuildTags = goConfig['buildTags'] ? goConfig['buildTags'].split(' ') : null;
+	let fileBuildTags = [];
+	const openSettings = {title: 'Open settings'};
+
+	for (let i = 0; i < e.document.lineCount; i++) {
+		if (e.document.lineAt(i).isEmptyOrWhitespace) {
+			break;
+		}
+		let foundBuildTags = e.document.lineAt(i).text.match(/^\/\/\s*\+build\s+(.+)$/)[1].split(' ');
```
Comment (ramya-rao-a) [inline src/goCheck.ts]: When there is no match, the `match` function returns `null`. Accessing the array member on `null` will throw error

File: src/goCheck.ts [hunk 5]
```diff
@@ -45,6 +45,37 @@ export function notifyIfGeneratedFile(e: vscode.TextDocumentChangeEvent) {
...
+		if (foundBuildTags) {
+			foundBuildTags = foundBuildTags.filter(item => item !== '');
+			fileBuildTags.push(...foundBuildTags);
+		}
+	}
+
+	if (fileBuildTags) {
+		for (let i = 0; i < fileBuildTags.length; i++) {
```
Comment (ramya-rao-a) [inline src/goCheck.ts]: Take the example of `// +build darwin freebsd netbsd openbsd`
This code here will show the prompt if any of the above is not present in the settings.
But according to https://dave.cheney.net/2013/10/12/how-to-use-conditional-compilation-with-the-go-build-tool,  "a build tag is evaluated as the OR of space-separated options"

So having any one of the above in the build tags is a sufficient.

Also, reading on in the same article, we have , "each option evaluates as the AND of its comma-separated terms" and "A file may have multiple build tags. The overall constraint is the logical AND of the individual constraints"

So looks like a lot more logic needs to be invloved here

Comment (urohit011) [reply src/goCheck.ts]: Okay, I get it, I only have a one question:
If the file has build tags like 
```// +build linux```
```// +build 386```, should it be present in the settings file like
 ```goBuildTags: linux,darwin``` as the goBuildTags holds a string?

Comment (ramya-rao-a) [reply src/goCheck.ts]: @urohit011 The `go.buildTags` takes space separated words.

So in your example the setting should be `"go.buildTags": "linux 386"`

If the file has `// +build abc def xyz`, then the build tags setting should contain either `abc`, `def` or `xyz`

Comment (urohit011) [reply src/goCheck.ts]: Hello @ramya-rao-a, thanks for the explanation, In my question what I actually wanted to know was  what should the ```"go.buildTags"``` should look like if the file contains two separate lines such as:
```// +build linux```
```// +build 386```

Comment (ramya-rao-a) [reply src/goCheck.ts]: It should look like this: `"go.buildTags": "linux 386"`

Comment (urohit011) [reply src/goCheck.ts]: Got it @ramya-rao-a , working on it

**Comment by ramya-rao-a** [comment]
Closing this PR as it hasnt seen any activity in over an year.
The general approach here is right, if anyone wants to pick it up and create a new PR, they are most welcome

**Pull request closed without merging**